src/onnx/parse_if.cpp

```diff
@@ -65,13 +67,108 @@ struct parse_if : op_parser<parse_if>
 
         auto then_out_shapes = then_mdl->get_output_shapes();
         auto else_out_shapes = else_mdl->get_output_shapes();
-        if(not std::equal(then_out_shapes.begin(),
-                          then_out_shapes.end(),
-                          else_out_shapes.begin(),
-                          else_out_shapes.end()))
-        {
+
+        auto throw_shapes = [&]() {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_graphs must have compatible shapes ");
+        };
+
+        if(then_out_shapes.size() != else_out_shapes.size())
+        {
+            throw_shapes();
+        }
+
+        // Add checks for each output shape
+        for(int i = 0; i < then_out_shapes.size(); i++)
+        {
+            // Must have the same type for both if/else blocks by onnx spec
+            if(then_out_shapes.at(i).type() != else_out_shapes.at(i).type())
+            {
+                MIGRAPHX_THROW("PARSE_IF: " + info.name +
+                               " then and else sub_grahps must have same output type! " +
+                               then_out_shapes.at(i).type_string() + " vs " +
+                               else_out_shapes.at(i).type_string());
+            }
+
+            if(not then_out_shapes.at(i).dynamic() && not else_out_shapes.at(i).dynamic())
+            {
+                auto then_lens = then_out_shapes.at(i).lens();
+                auto else_lens = else_out_shapes.at(i).lens();
+
+                // Throw error if both branches have zero output shapes. Not possible for static
+                // inputs
+                if(then_lens.empty() && else_lens.empty())
+                {
+                    throw_shapes();
+                }
+
+                auto handle_empty_branch = [](module_ref& mdl, const shape& out_shape) {
+                    auto outline_ins = mdl->add_outline(out_shape);
```

Comment:
Dont use `add_outline`. This will throw an error if the empty branch is picked. I also plan to remove `outline` in the future since its very problematic and is never helpful.

Reply:
In this case it does in the input is an empty constant shape to one of the branches, otherwise we'll fail on the if_op of shapes being invalid. I'm using that to gate the size.

Reply:
> empty constant shape to one of the branches

What does this mean?

Reply:
Somehow the model generated had inputs to one of the then/else branches as an empty shape constant 0 with a type and then  specific type for the other part of the corresponding IF output. We have no way of determining the shape of that empty constant, only its type. By adding the outline using the other branch's shape is within the onnx spec for how those scenarios are handled.

Reply:
What is an empty constant?

Reply:
One of the the inputs to the IF node in the onnx converted model of rexnext50. There isn't shape information outside an empty shape, and a type to use the other branch and that's what fails when running through that model.

Reply:
![image](https://user-images.githubusercontent.com/107195283/196323325-9992823e-af95-4b25-8820-2cee72554095.png)

This part of the Onnx model was added to one of the IF blocks and was a giant source of pain is what I mean. Originally this was being parsed in by default to as a float until @CharlieL7 added in a change to shape() constructor since we kept failing on type. Now the issue is how to infer shape, which there is none.

Without the outline copied from the other model I'm not sure how best to add the proper shape from the other branch. This is valid Onnx as this "empty" shape, but identical type to the other, non empty branch and the output should be the union of those.

Unless the better option is to just create a new literal with matching shapes and then remove the outline entirely?

Reply:
But does this constant have a value?

I understand an "empty" shape would mean it was a scalar, but an empty constant would not be a valid ONNX, see from [`Constant`](https://github.com/onnx/onnx/blob/main/docs/Operators.md#constant) doc:

> Exactly one of the provided attributes, either value, sparse_value, or value_* must be specified.

And from there we should be able to unsqueeze the scalar to match the shape of the other branch.

Reply:
Well the unsqueeze doesn't seem to work in this instance. I've tried it and got errors with 

```UNSQUEZE: Input must be scalar.```

I've added a change that removes the use of outline and instead comes up with a new literal with the correct size instead.

Reply:
Ah wait, `value` attribute is for a tensor value, but this is a scalar. Either way we should have at least one value. ONNX requires at least one value, from [Tensor Definition](https://github.com/onnx/onnx/blob/main/docs/IR.md#tensor-definition)(where `S` is the shape and `V` is the values):

> If S has length 0, V must have length 1, since the empty product is defined to be 1. In this case, the tensor represents a scalar.

Reply:
Does that mean squeeze is broken?

```
        if(input_shape.scalar())
        {
            if(old_lens.size() == 1 and old_lens.front() == 1)
                return shape{type, old_lens};
            else
                MIGRAPHX_THROW("UNSQUEEZE: Input must be a scalar");
        }
```
Is where I seem to be failing when I use unsqueeze.

Reply:
Why would it fail there? If the dimensions are empty then we should have a shape of `{1}, {0}`. It seems like we arent passing the shape correctly.

Reply:
Can you show me what this branch looks like in migraphx IR?

Reply:
I think its due to how shape is handling empty inputs right now it defaults to the following

```
 /*!
     * Empty literal with a specific shape type
     */
    explicit literal(shape::type_t shape_type) : m_shape(shape_type, {}) {}
```

Should this be just 

```
 /*!
     * Empty literal with a specific shape type
     */
    explicit literal(shape::type_t shape_type) : m_shape(shape_type, {1}) {}
```
So 1, with no stride instead?

Reply:
```
[   RUN    ] if_then_empty_constant_test
If_4_if:@0 = @literal{2, 1} -> int64_type, {2, 1}, {1, 1}
If_4_if:@1 = @literal{} -> int64_type, {}, {}
If_4_if:@2 = @return(If_4_if:@0)

If_4_else:@0 = mul(?,?) -> int64_type, {2, 1}, {1, 1}
If_4_else:@1 = @return(If_4_else:@0)
```

Using MIGraphX IR

Reply:
> Trying with both zero len and stride seems toalso break if_literal_test since I assume we are expecting to get shape_type {}, {} for empty values?

Yea right now we dont represent scalars as empty dimensions, we have #519 to update this. Right now we represent the scalar as `{1}, {0}`.

Reply:
So `@literal{} -> int64_type, {}, {}` doesnt look like it was parsed correctly. A literal from onnx should always have a value. If the dimensions are empty in onnx it should look like this `@literal{5} -> int64_type, {1}, {0}`(assuming it has a scalar value of `5`). 

Since the shape has to go to `{2, 1}` then we will need to do a broadcast and a contiguous. We should probably call `common_shape` to get the common shape and then insert multibroadcasts+contiguous to get the common shape. And we could probably do this for all cases.

Reply:
> Should I be adding that as part of this PR now too?

No.

> Well I'd assume here then we can just say an empty value = 0? so it would resolve to

The value can't be empty. Onnx spec requires a value.